@wdio/utils:shim: SyntaxError: Unexpected token Y in JSON at position 0

**Environment (please complete the following information):**
 - **WebdriverIO version:** 7.9.1
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** async
 - **Node.js version:** 14.17.6
 - **NPM version:** 6.14.15
 - **Browser name and version:** Chrome 91
 - **Platform name and version:** Mac OS Big Sur
 - **Additional wdio packages used (if applicable):** @wdio/devtools-service v7.9.1

**Config of WebdriverIO**
```
    ...
    hostname: '10.0.x.y',
    port: 4444,
    path: '/wd/hub',
    ....
    // Test runner services
    // Services take over a specific job you don't want to take care of. They enhance
    // your test setup with almost no effort. Unlike plugins, they don't add new
    // commands. Instead, they hook themselves up into the test process.
    services: ['devtools', 'rerun'],
    ...
    capabilities: [
    {
        browserName: 'chrome',
        'selenoid:options': { enableVNC: true, enableVideo: false, hostsEntries },
    },
    ],
    ...
],
```

**Describe the bug**
After the `devtool` service activation, error messages like this:
```
[0-0] 2021-10-08T07:49:55.455Z ERROR @wdio/utils:shim: SyntaxError: Unexpected token Y in JSON at position 0
[0-0]     at JSON.parse (<anonymous>)
[0-0]     at IncomingMessage.<anonymous> (/***************/e2e-wd/node_modules/lighthouse/lighthouse-core/gather/connections/cri.js:97:28)
[0-0]     at IncomingMessage.emit (events.js:412:35)
[0-0]     at IncomingMessage.emit (domain.js:470:12)
[0-0]     at endReadableNT (internal/streams/readable.js:1317:12)
[0-0]     at processTicksAndRejections (internal/process/task_queues.js:82:21)
[0-0] Error in "AfterHook Hook"
Unexpected token Y in JSON at position 0
```
started to appear in the output. I checked the actual request and it indeed attempts to read JSON at `http://10.0.x.y:4444/json/list`. The problem is that this endpoint responds with plaintext `You are using Selenoid 1.10.5!`.

It does not affect the execution of the tests though.

**To Reproduce**
Steps to reproduce the behavior:

* Use wdio with almost default configuration with `hostname` pointing to a [Selenoid](https://aerokube.com/selenoid/latest/) cluster entry point.

**Expected behavior**
No error messages should appear in the logs.

**Log**

See an example error message above.

Your example has the same issue when executed against my Selenoid instance:
```
(base) dmitry.lomakin@dlomakin-mb wdio-selenoid-examples % npm test

> wdio-selenoid-examples@0.0.1 test /Users/dmitry.lomakin/Sources/wdio-selenoid-examples
> wdio


Execution of 1 workers started at 2021-10-08T08:29:15.294Z

2021-10-08T08:29:15.745Z INFO @wdio/cli:launcher: Run onPrepare hook
2021-10-08T08:29:15.747Z INFO @wdio/cli:launcher: Run onWorkerStart hook
2021-10-08T08:29:15.748Z INFO @wdio/local-runner: Start worker 0-0 with arg:
[0-0] 2021-10-08T08:29:16.204Z INFO @wdio/local-runner: Run worker command: run
[0-0] RUNNING in chrome - /test/specs/example.e2e.ts
[0-0] 2021-10-08T08:29:17.044Z INFO webdriver: Initiate new session using the WebDriver protocol
[0-0] 2021-10-08T08:29:17.092Z INFO webdriver: [POST] http://10.0.X.Y:4444/wd/hub/session
[0-0] 2021-10-08T08:29:17.092Z INFO webdriver: DATA {
[0-0]   capabilities: {
[0-0]     alwaysMatch: {
[0-0]       browserName: 'chrome',
[0-0]       acceptInsecureCerts: true,
[0-0]       'selenoid:options': [Object]
[0-0]     },
[0-0]     firstMatch: [ {} ]
[0-0]   },
[0-0]   desiredCapabilities: {
[0-0]     browserName: 'chrome',
[0-0]     acceptInsecureCerts: true,
[0-0]     'selenoid:options': { enableVNC: true }
[0-0]   }
[0-0] }
[0-0] 2021-10-08T08:29:19.579Z ERROR @wdio/utils:shim: SyntaxError: Unexpected token Y in JSON at position 0
[0-0]     at JSON.parse (<anonymous>)
[0-0]     at IncomingMessage.<anonymous> (/Users/dmitry.lomakin/Sources/wdio-selenoid-examples/node_modules/lighthouse/lighthouse-core/gather/connections/cri.js:97:28)
[0-0]     at IncomingMessage.emit (events.js:412:35)
[0-0]     at IncomingMessage.emit (domain.js:470:12)
[0-0]     at endReadableNT (internal/streams/readable.js:1317:12)
[0-0]     at processTicksAndRejections (internal/process/task_queues.js:82:21)
[0-0] 2021-10-08T08:29:19.585Z INFO webdriver: COMMAND navigateTo("https://webdriver.io/")
[0-0] 2021-10-08T08:29:19.585Z INFO webdriver: [POST] http://10.0.X.Y:4444/wd/hub/session/1b42cee1833f22e9622fbe977db8369d/url
[0-0] 2021-10-08T08:29:19.585Z INFO webdriver: DATA { url: 'https://webdriver.io/' }
[0-0] {
[0-0]   Timestamp: 2589521.022378,
[0-0]   Documents: 25,
[0-0]   Frames: 19,
[0-0]   JSEventListeners: 300,
[0-0]   Nodes: 2143,
[0-0]   LayoutCount: 10,
[0-0]   RecalcStyleCount: 22,
[0-0]   LayoutDuration: 0.189904,
[0-0]   RecalcStyleDuration: 0.08845,
[0-0]   ScriptDuration: 0.430856,
[0-0]   TaskDuration: 1.390678,
[0-0]   JSHeapUsedSize: 5730368,
[0-0]   JSHeapTotalSize: 9658368
[0-0] }
[0-0] 2021-10-08T08:29:21.225Z INFO webdriver: COMMAND getWindowHandle()
[0-0] 2021-10-08T08:29:21.225Z INFO webdriver: [GET] http://10.0.X.Y:4444/wd/hub/session/1b42cee1833f22e9622fbe977db8369d/window
[0-0] 2021-10-08T08:29:21.283Z INFO webdriver: RESULT CDwindow-49C55A9B76B9BBC4BE8AAD259A4058D6
[0-0] 2021-10-08T08:29:21.454Z INFO webdriver: COMMAND navigateTo("https://todobackend.com/client/index.html?https://todo-backend-spring4-java8.herokuapp.com/todos/")
[0-0] 2021-10-08T08:29:21.454Z INFO webdriver: [POST] http://10.0.X.Y:4444/wd/hub/session/1b42cee1833f22e9622fbe977db8369d/url
[0-0] 2021-10-08T08:29:21.454Z INFO webdriver: DATA {
[0-0]   url: 'https://todobackend.com/client/index.html?https://todo-backend-spring4-java8.herokuapp.com/todos/'
[0-0] }
[0-0] Error in "devtools.should allow use mock feature"
Error: Expect mock to be called 1 time

Expected: 1
Received: 0
    at Context.<anonymous> (/Users/dmitry.lomakin/Sources/wdio-selenoid-examples/test/specs/example.e2e.ts:26:28)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
[0-0] 2021-10-08T08:29:27.562Z INFO webdriver: COMMAND deleteSession()
[0-0] 2021-10-08T08:29:27.562Z INFO webdriver: [DELETE] http://10.0.X.Y:4444/wd/hub/session/1b42cee1833f22e9622fbe977db8369d
[0-0] FAILED in chrome - /test/specs/example.e2e.ts
2021-10-08T08:29:27.806Z INFO @wdio/cli:launcher: Run onComplete hook

 "spec" Reporter:
------------------------------------------------------------------
[chrome 92.0.4515.131 linux #0-0] Running: chrome (v92.0.4515.131) on linux
[chrome 92.0.4515.131 linux #0-0] Session ID: 1b42cee1833f22e9622fbe977db8369d
[chrome 92.0.4515.131 linux #0-0]
[chrome 92.0.4515.131 linux #0-0] » /test/specs/example.e2e.ts
[chrome 92.0.4515.131 linux #0-0] devtools
[chrome 92.0.4515.131 linux #0-0]    ✓ should allow use puppeteer instance
[chrome 92.0.4515.131 linux #0-0]    ✖ should allow use mock feature
[chrome 92.0.4515.131 linux #0-0]
[chrome 92.0.4515.131 linux #0-0] 1 passing (8.1s)
[chrome 92.0.4515.131 linux #0-0] 1 failing
[chrome 92.0.4515.131 linux #0-0]
[chrome 92.0.4515.131 linux #0-0] 1) devtools should allow use mock feature
[chrome 92.0.4515.131 linux #0-0] Expect mock to be called 1 time

Expected: 1
Received: 0
[chrome 92.0.4515.131 linux #0-0] Error: Expect mock to be called 1 time
[chrome 92.0.4515.131 linux #0-0]
[chrome 92.0.4515.131 linux #0-0] Expected: 1
[chrome 92.0.4515.131 linux #0-0] Received: 0
[chrome 92.0.4515.131 linux #0-0]     at Context.<anonymous> (/Users/dmitry.lomakin/Sources/wdio-selenoid-examples/test/specs/example.e2e.ts:26:28)
[chrome 92.0.4515.131 linux #0-0]     at processTicksAndRejections (internal/process/task_queues.js:95:5)


Spec Files:	 0 passed, 1 failed, 1 total (100% completed) in 00:00:12

2021-10-08T08:29:27.807Z INFO @wdio/local-runner: Shutting down spawned worker
2021-10-08T08:29:28.058Z INFO @wdio/local-runner: Waiting for 0 to shut down gracefully
2021-10-08T08:29:28.059Z INFO @wdio/local-runner: shutting down
npm ERR! Test failed.  See above for more details.
```

--- a/packages/wdio-devtools-service/src/index.ts
+++ b/packages/wdio-devtools-service/src/index.ts
@@ -238,7 +238,7 @@ export default class DevToolsService implements Services.ServiceInstance {
         }
 
         this._session = await this._target.createCDPSession()
-        this._driver = await getLighthouseDriver(this._session)
+        this._driver = await getLighthouseDriver(this._session, this._target)
 
         new CommandHandler(this._session, this._page, this._browser)
         this._traceGatherer = new TraceGatherer(this._session, this._page, this._driver)

--- a/packages/wdio-devtools-service/src/lighthouse/cri.ts
+++ b/packages/wdio-devtools-service/src/lighthouse/cri.ts
@@ -1,0 +1,34 @@
+import CriConnection from 'lighthouse/lighthouse-core/gather/connections/cri'
+
+const DEFAULT_HOSTNAME = 'localhost'
+const DEFAULT_PORT = '9222'
+
+/**
+ * this class got patched to enable connecting to a remote path like
+ * ws://192.168.0.39:4444/session/349a44a32846c2659c703e71403bd472/se/cdp
+ * as it requires to attach to a session before.
+ */
+export default class ChromeProtocolPatched extends CriConnection {
+    private _sessionId?: string
+
+    /**
+     * Add constructor for typing safety
+     * @param {number=} port Optional port number. Defaults to 9222;
+     * @param {string=} hostname Optional hostname. Defaults to localhost.
+     * @constructor
+     */
+    constructor(port: string = DEFAULT_PORT, hostname: string = DEFAULT_HOSTNAME) {
+        super(port, hostname)
+    }
+
+    setSessionId (sessionId: string) {
+        this._sessionId = sessionId
+    }
+
+    /**
+     * force every command to be send with the given session id
+     */
+    sendCommand(method: string, sessionId?: string, ...paramArgs: any[]) {
+        return super.sendCommand(method, sessionId || this._sessionId, ...paramArgs)
+    }
+}

--- a/packages/wdio-devtools-service/src/utils.ts
+++ b/packages/wdio-devtools-service/src/utils.ts
@@ -1,9 +1,10 @@
 import type { Browser, MultiRemoteBrowser } from 'webdriverio'
 import type { Capabilities } from '@wdio/types'
-import ChromeProtocol from 'lighthouse/lighthouse-core/gather/connections/cri'
 import Driver from 'lighthouse/lighthouse-core/gather/driver'
 import type { CDPSession } from 'puppeteer-core/lib/cjs/puppeteer/common/Connection'
+import type { Target } from 'puppeteer-core/lib/cjs/puppeteer/common/Target'
 
+import ChromeProtocol from './lighthouse/cri'
 import { IGNORED_URLS, UNSUPPORTED_ERROR_MESSAGE } from './constants'
 import { RequestPayload } from './handler/network'
 import type { GathererDriver } from './types'
@@ -88,10 +89,29 @@ export function isBrowserSupported(caps: Capabilities.Capabilities) {
     return true
 }
 
-export async function getLighthouseDriver (session: CDPSession): Promise<GathererDriver> {
+/**
+ * Either request the page list directly from the browser or if Selenium
+ * or Selenoid is used connect to a target manually
+ */
+export async function getLighthouseDriver (session: CDPSession, target: Target): Promise<GathererDriver> {
     const c = session.connection().url()
     const cUrl = new URL(c)
     const connection = new ChromeProtocol(cUrl.port, cUrl.hostname)
+
+    if (cUrl.protocol === 'ws:') {
+        await connection._connectToSocket({
+            webSocketDebuggerUrl: c,
+            id: target._targetId
+        })
+        const { sessionId } = await connection.sendCommand(
+            'Target.attachToTarget',
+            undefined,
+            { targetId: target._targetId, flatten: true }
+        )
+        connection.setSessionId(sessionId)
+        return new Driver(connection)
+    }
+
     const list = await connection._runJsonCommand('list')
     await connection._connectToSocket(list[0])
     return new Driver(connection)

--- a/packages/wdio-devtools-service/src/utils.ts
+++ b/packages/wdio-devtools-service/src/utils.ts
@@ -98,7 +98,11 @@ export async function getLighthouseDriver (session: CDPSession, target: Target):
     const cUrl = new URL(c)
     const connection = new ChromeProtocol(cUrl.port, cUrl.hostname)
 
-    if (cUrl.protocol === 'ws:') {
+    /**
+     * only create a new DevTools session if our WebSocket url doesn't already indicate
+     * that we are using one
+     */
+    if (!cUrl.pathname.startsWith('/devtools/browser')) {
         await connection._connectToSocket({
             webSocketDebuggerUrl: c,
             id: target._targetId

--- a/packages/wdio-devtools-service/tests/utils.test.ts
+++ b/packages/wdio-devtools-service/tests/utils.test.ts
@@ -13,6 +13,17 @@ jest.mock('fs', () => ({
     existsSync: jest.fn()
 }))
 
+jest.mock('lighthouse/lighthouse-core/gather/connections/cri', () => class ChromeProtocol {
+    public _runJsonCommand = jest.fn().mockReturnValue(['foobar'])
+    public _connectToSocket = jest.fn()
+    public _sendCommandMock = jest.fn()
+
+    sendCommand (...args) {
+        this._sendCommandMock(...args)
+        return { sessionId: 'session 321' }
+    }
+})
+
 test('sumByKey', () => {
     expect(sumByKey([{
         size: 1
@@ -110,14 +121,28 @@ describe('isBrowserSupported', () => {
 })
 
 describe('getLighthouseDriver', () => {
-    test('should return a driver', async () => {
+    test('should return a driver w/o creating new session', async () => {
         const urlMock = jest.fn().mockReturnValue('ws://127.0.0.1:56513/devtools/browser/9aae0e34-86a9-4b0e-856b-d0d37009ddbb')
         const session = {
             connection: jest.fn().mockReturnValue({ url: urlMock })
         }
-        const driver = await getLighthouseDriver(session as any)
+        const target = { _targetId: 'foobar321' }
+        const driver = await getLighthouseDriver(session as any, target as any)
         expect(session.connection).toBeCalledTimes(1)
         expect(urlMock).toBeCalledTimes(1)
         expect(driver.constructor.name).toBe('Driver')
+    })
+
+    test('should create a new session', async () => {
+        const urlMock = jest.fn().mockReturnValue('ws://127.0.0.1:56513/session/9aae0e34-86a9-4b0e-856b-d0d37009ddbb/se/cdp')
+        const session = {
+            connection: jest.fn().mockReturnValue({ url: urlMock })
+        }
+        const target = { _targetId: 'foobar321' }
+        const driver = await getLighthouseDriver(session as any, target as any)
+        // @ts-expect-error
+        driver.connection.sendCommand('foobar')
+        // @ts-expect-error
+        expect(driver.connection._sendCommandMock.mock.calls).toMatchSnapshot()
     })
 })

